utils/smartpaste/mattermost.tsx

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Options} from 'turndown';
+
+export const channelMentionsRule = {
+    filter: (node: Node): boolean => {
+        return node.nodeName === 'A' && (node as HTMLElement).className === 'mention-link' && Boolean((node as HTMLElement).getAttribute('data-channel-mention'));
+    },
+    replacement: (content: string, node: Node): string => {
+        const channelName = (node as HTMLElement).getAttribute('data-channel-mention');
+        if (channelName) {
+            return `~${channelName}`;
+        }
+        return '';
+    }
+};
+
+export const hashtagsRule = {
+    filter: (node: Node): boolean => {
+        return node.nodeName === 'A' && (node as HTMLElement).className === 'mention-link' && Boolean((node as HTMLElement).getAttribute('data-hashtag'));
+    },
+    replacement: (content: string, node: Node): string => {
+        return (node as HTMLElement).getAttribute('data-hashtag') || '';
+    }
+};
+
+export const filePreviewButtonRule = {
+    filter: (node: Node): boolean => {
+        return node.nodeName === 'DIV' && (node as HTMLElement).className === 'file-preview__button';
+    },
+    replacement: (content: string): string => {
+        return content;
+    }
+};
+
+export const codeBlockRule = {
+    filter: (node: Node) => {
+        return node.nodeName === 'CODE' && (node as HTMLElement).className.indexOf('hljs') !== -1;
+    },
+
+    replacement: (content: string, node: Node, options: Options) => {
+        const language = ((node as HTMLElement).className.match(/language-(\S+)/) || [null, ''])[1];
+        const code = node.textContent || '';
+
+        const fenceChar = options.fence?.charAt(0) || '`';
```

Comment:
Do we need different handling for when a code block is fenced with a ~ instead of a backtick? You might need that if, for example, you're quoting code containing triple backticks like

    ~~~
    Here's how to use a code block:
    ```java
    public static void main(String[] args) {
        System.out.Println("Hello, World!");
    }
    ```
    ~~~

Reply:
Oh, I think I see. This code is handling that by increasing the size of the fence? I didn't actually know that was possible. I thought you had to do stuff with tildes or indented code blocks for that to happen.

Reply:
Yes, exactly.